File: frontend/js/file-upload.js
const API_BASE = "/api/v1";

const DEFAULT_LIMITS = {
  max_file_size_bytes: 10 * 1024 * 1024,
  max_files_per_message: 3,
  chunk_size_bytes: 1024 * 1024,
  allowed_extensions: [".jpg", ".jpeg", ".png", ".pdf", ".txt"],
  allowed_content_types: [
    "image/jpeg",
    "image/jpg",
    "image/png",
    "application/pdf",
    "text/plain",
  ],
};

export class FileUploadManager {
  constructor(session) {
    this.session = session;
    this.limits = null;
    this.pending = [];
  }

  async getLimits() {
    if (this.limits) return this.limits;
    try {
      const res = await fetch(`${API_BASE}/chat/uploads/limits`);
      if (res.ok) {
        this.limits = await res.json();
        return this.limits;
      }
    } catch (err) {
      console.warn("Using default upload limits", err);
    }
    this.limits = DEFAULT_LIMITS;
    return this.limits;
  }

  validateFile(file, limits, currentCount) {
    if (currentCount >= limits.max_files_per_message) {
      return `Maximum ${limits.max_files_per_message} files per message`;
    }
    if (file.size > limits.max_file_size_bytes) {
      return `File must be under ${limits.max_file_size_bytes / (1024 * 1024)}MB`;
    }
    const ext = `.${file.name.split(".").pop()?.toLowerCase() || ""}`;
    if (!limits.allowed_extensions.includes(ext)) {
      return `File type not allowed (${ext})`;
    }
    const type = file.type || "application/octet-stream";
    if (!limits.allowed_content_types.includes(type)) {
      return `Content type not allowed (${type})`;
    }
    return null;
  }

  async uploadFile(file, onProgress) {
    const limits = await this.getLimits();
    const initRes = await fetch(`${API_BASE}/chat/uploads/init`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        anonymous_user_id: this.session.anonymous_user_id,
        conversation_id: this.session.conversation_id,
        filename: file.name,
        content_type: file.type || "application/octet-stream",
        size_bytes: file.size,
      }),
    });
    if (!initRes.ok) {
      const err = await initRes.json().catch(() => ({}));
      throw new Error(err.detail || `Upload init failed (${initRes.status})`);
    }
    const init = await initRes.json();
    const chunkSize = init.chunk_size;
    let uploaded = 0;

    for (const chunk of init.chunks) {
      const start = chunk.chunk_index * chunkSize;
      const end = Math.min(start + chunkSize, file.size);
      const blob = file.slice(start, end);
      let url = chunk.upload_url;
      if (init.storage_provider === "local") {
        const sep = url.includes("?") ? "&" : "?";
        url = `${url}${sep}anonymous_user_id=${this.session.anonymous_user_id}`;
      }
      if (!url.startsWith("http")) {
        url = `${window.location.origin}${url}`;
      }
      const putRes = await fetch(url, {
        method: chunk.method || "PUT",
        body: blob,
        headers: { "Content-Type": file.type || "application/octet-stream" },
      });
      if (!putRes.ok) {
        throw new Error(`Chunk ${chunk.chunk_index + 1} upload failed`);
      }
      uploaded = end;
      if (onProgress) {
        onProgress(Math.round((uploaded / file.size) * 100));
      }
    }

    const completeRes = await fetch(
      `${API_BASE}/chat/uploads/${init.upload_id}/complete`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          anonymous_user_id: this.session.anonymous_user_id,
          conversation_id: this.session.conversation_id,
          upload_id: init.upload_id,
        }),
      },
    );
    if (!completeRes.ok) {
      const err = await completeRes.json().catch(() => ({}));
      throw new Error(err.detail || `Upload complete failed (${completeRes.status})`);
    }
    return completeRes.json();
  }

  createPreview(file, uploaded = null) {
    const item = {
      id: crypto.randomUUID(),
      file,
      uploaded,
      previewUrl: null,
      progress: uploaded ? 100 : 0,
      error: null,
    };
    if (file.type.startsWith("image/")) {
      item.previewUrl = URL.createObjectURL(file);
    } else if (uploaded?.preview_url) {
      item.previewUrl = uploaded.preview_url;
    }
    return item;
  }

  revokePreview(item) {
    if (item.previewUrl?.startsWith("blob:")) {
      URL.revokeObjectURL(item.previewUrl);
    }
  }
}

File: backend/app/services/file_upload.py
import json
import math
from pathlib import Path
from uuid import UUID, uuid4

from app.core.config import settings
from app.core.redis import get_redis_client
from app.schemas.upload import (
    AttachmentType,
    ChunkUploadInfo,
    CompleteUploadResponse,
    InitUploadRequest,
    InitUploadResponse,
    UploadLimitsResponse,
)
from app.services.storage.factory import get_storage_backend

UPLOAD_KEY_PREFIX = "upload"

ALLOWED_CONTENT_TYPES: dict[str, AttachmentType] = {
    "image/jpeg": AttachmentType.image,
    "image/jpg": AttachmentType.image,
    "image/png": AttachmentType.image,
    "application/pdf": AttachmentType.pdf,
    "text/plain": AttachmentType.text,
}

ALLOWED_EXTENSIONS = {".jpg", ".jpeg", ".png", ".pdf", ".txt"}


class FileUploadError(ValueError):
    pass


class FileUploadService:
    def __init__(self) -> None:
        self._redis = get_redis_client()
        self._storage = get_storage_backend()

    def limits(self) -> UploadLimitsResponse:
        return UploadLimitsResponse(
            max_file_size_bytes=settings.upload_max_file_size_bytes,
            max_files_per_message=settings.upload_max_files_per_message,
            chunk_size_bytes=settings.upload_chunk_size_bytes,
            allowed_extensions=sorted(ALLOWED_EXTENSIONS),
            allowed_content_types=sorted(ALLOWED_CONTENT_TYPES.keys()),
        )

    def _upload_key(self, upload_id: UUID) -> str:
        return f"{UPLOAD_KEY_PREFIX}:{upload_id}"

    def _attachment_type(self, content_type: str, filename: str) -> AttachmentType:
        normalized = content_type.lower().split(";")[0].strip()
        if normalized in ALLOWED_CONTENT_TYPES:
            return ALLOWED_CONTENT_TYPES[normalized]
        ext = Path(filename).suffix.lower()
        if ext in {".jpg", ".jpeg", ".png"}:
            return AttachmentType.image
        if ext == ".pdf":
            return AttachmentType.pdf
        if ext == ".txt":
            return AttachmentType.text
        raise FileUploadError(f"File type not allowed: {content_type}")

    def validate_file(self, *, filename: str, content_type: str, size_bytes: int) -> None:
        if size_bytes <= 0:
            raise FileUploadError("File is empty")
        if size_bytes > settings.upload_max_file_size_bytes:
            raise FileUploadError(
                f"File exceeds {settings.upload_max_file_size_bytes // (1024 * 1024)}MB limit"
            )
        ext = Path(filename).suffix.lower()
        if ext not in ALLOWED_EXTENSIONS:
            raise FileUploadError(
                f"Extension '{ext}' not allowed. Allowed: {', '.join(sorted(ALLOWED_EXTENSIONS))}"
            )
        normalized = content_type.lower().split(";")[0].strip()
        if normalized not in ALLOWED_CONTENT_TYPES:
            raise FileUploadError(
                f"Content type '{content_type}' not allowed"
            )
        self._attachment_type(content_type, filename)

    async def init_upload(self, request: InitUploadRequest) -> InitUploadResponse:
        self.validate_file(
            filename=request.filename,
            content_type=request.content_type,
            size_bytes=request.size_bytes,
        )

        upload_id = uuid4()
        file_id = uuid4()
        chunk_size = settings.upload_chunk_size_bytes
        total_chunks = max(1, math.ceil(request.size_bytes / chunk_size))

        chunks = await self._storage.create_chunk_targets(
            upload_id=upload_id,
            file_id=file_id,
            filename=request.filename,
            content_type=request.content_type,
            total_chunks=total_chunks,
        )

        session = {
            "upload_id": str(upload_id),
            "file_id": str(file_id),
            "filename": request.filename,
            "content_type": request.content_type,
            "size_bytes": request.size_bytes,
            "total_chunks": total_chunks,
            "conversation_id": str(request.conversation_id),
            "anonymous_user_id": str(request.anonymous_user_id),
        }
        await self._redis.set(
            self._upload_key(upload_id),
            json.dumps(session),
            ex=settings.upload_signed_url_ttl_seconds,
        )

        return InitUploadResponse(
            upload_id=upload_id,
            file_id=file_id,
            chunk_size=chunk_size,
            total_chunks=total_chunks,
            chunks=[
                ChunkUploadInfo(
                    chunk_index=c.chunk_index,
                    upload_url=c.upload_url,
                    method=c.method,
                )
                for c in chunks
            ],
            storage_provider=settings.storage_provider,
        )

    async def _get_session(self, upload_id: UUID) -> dict:
        raw = await self._redis.get(self._upload_key(upload_id))
        if not raw:
            raise FileUploadError("Upload session expired or not found")
        return json.loads(raw)

    async def store_chunk(
        self,
        upload_id: UUID,
        chunk_index: int,
        data: bytes,
        *,
        anonymous_user_id: UUID,
    ) -> None:
        session = await self._get_session(upload_id)
        if session["anonymous_user_id"] != str(anonymous_user_id):
            raise FileUploadError("Upload session does not belong to this user")
        if chunk_index < 0 or chunk_index >= session["total_chunks"]:
            raise FileUploadError("Invalid chunk index")
        await self._storage.store_chunk(
            upload_id=upload_id,
            chunk_index=chunk_index,
            data=data,
        )

    async def complete_upload(
        self,
        upload_id: UUID,
        *,
        anonymous_user_id: UUID,
        conversation_id: UUID,
    ) -> CompleteUploadResponse:
        session = await self._get_session(upload_id)
        if session["anonymous_user_id"] != str(anonymous_user_id):
            raise FileUploadError("Upload session does not belong to this user")
        if session["conversation_id"] != str(conversation_id):
            raise FileUploadError("Upload session does not belong to this conversation")

        file_id = UUID(session["file_id"])
        stored = await self._storage.finalize_upload(
            upload_id=upload_id,
            file_id=file_id,
            filename=session["filename"],
            content_type=session["content_type"],
            total_chunks=session["total_chunks"],
        )
        await self._redis.delete(self._upload_key(upload_id))

        attachment_type = self._attachment_type(
            session["content_type"],
            session["filename"],
        )
        return CompleteUploadResponse(
            file_id=stored.file_id,
            url=stored.public_url,
            filename=session["filename"],
            content_type=session["content_type"],
            attachment_type=attachment_type,
            size_bytes=session["size_bytes"],
            preview_url=stored.preview_url,
        )
